tests/kernel/ipi_optimize/src/main.c

```diff
@@ -0,0 +1,741 @@
+/*
+ * Copyright (c) 2024 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+#include <zephyr/kernel.h>
+#include <ksched.h>
+#include <zephyr/kernel_structs.h>
+
+#define STACK_SIZE (1024 + CONFIG_TEST_EXTRA_STACK_SIZE)
+
+#define NUM_THREADS (CONFIG_MP_MAX_NUM_CPUS - 1)
+
+static struct k_thread thread[NUM_THREADS];
+static struct k_thread alt_thread;
+
+static bool alt_thread_created;
+
+static K_THREAD_STACK_ARRAY_DEFINE(stack, NUM_THREADS, STACK_SIZE);
+static K_THREAD_STACK_DEFINE(alt_stack, STACK_SIZE);
+
+static uint32_t ipi_count[CONFIG_MP_MAX_NUM_CPUS];
+static struct k_spinlock ipilock;
+static atomic_t in_loop;
+
+static K_SEM_DEFINE(sem, 0, 1);
+
+void z_trace_sched_ipi(void)
+{
+	k_spinlock_key_t  key;
+
+	key = k_spin_lock(&ipilock);
+	ipi_count[_current_cpu->id]++;
+	k_spin_unlock(&ipilock, key);
```

Comment:
I did consider that. The reason I did not go that route was because when getting and clearing the IPI counts via get_ipi_counts() and clear_ipi_counts() respectively, I wanted to ensure that a snapshot of all of them were obtained at the same time. Atomics would be useful for each element, but unless I have missed something not for the whole array.